api/src/main/java/org/pmiops/workbench/db/dao/WorkspaceServiceImpl.java

```diff
@@ -0,0 +1,106 @@
+package org.pmiops.workbench.db.dao;
+
+import java.util.List;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.logging.Logger;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import org.pmiops.workbench.db.model.Workspace;
+import org.pmiops.workbench.db.model.WorkspaceUserRole;
+import org.pmiops.workbench.exceptions.BadRequestException;
+import org.pmiops.workbench.exceptions.NotFoundException;
+
+/**
+ * Workspace manipulation and shared business logic which can't be represented by automatic query
+ * generation in WorkspaceDao, or convenience aliases.
+ *
+ * TODO(RW-215) Add versioning to detect/prevent concurrent edits.
+ */
+@Service
+public class WorkspaceServiceImpl implements WorkspaceService {
+  private static final Logger log = Logger.getLogger(WorkspaceService.class.getName());
+
+  /**
+   * Clients wishing to use the auto-generated methods from the DAO interface may directly access
+   * it here.
+   */
+  @Autowired private WorkspaceDao workspaceDao;
+
+  public WorkspaceServiceImpl() {
+
+  }
+
+  public WorkspaceDao getDao() {
+    return workspaceDao;
+  }
+
+  public Workspace get(String ns, String id) {
+    return workspaceDao.findByWorkspaceNamespaceAndFirecloudName(ns, id);
+  }
+
+  public Workspace getRequired(String ns, String id) {
+    Workspace workspace = get(ns, id);
+    if (workspace == null) {
+      throw new NotFoundException(String.format("Workspace %s/%s not found.", ns, id));
+    }
+    return workspace;
+  }
+
+  public List<Workspace> findForReview() {
+    return workspaceDao.findByApprovedIsNullAndReviewRequestedTrueOrderByTimeRequested();
+  }
+
+  // FIXME @Version instead? Bean instantiation v. @Transactional?
+  //@Transactional
+  public void setResearchPurposeApproved(String ns, String id, boolean approved) {
+    Workspace workspace = getRequired(ns, id);
+    if (workspace.getReviewRequested() == null || !workspace.getReviewRequested()) {
+      throw new BadRequestException(String.format(
+          "No review requested for workspace %s/%s.", ns, id));
+    }
+    if (workspace.getApproved() != null) {
+      throw new BadRequestException(String.format(
+          "Workspace %s/%s already %s.",
+          ns, id, workspace.getApproved() ? "approved" : "rejected"));
+    }
+    workspace.setApproved(approved);
+    workspaceDao.save(workspace);
+  }
+
+  @Transactional
+  public void updateUserRoles(String ns, String id, Set<WorkspaceUserRole> userRoleSet) {
```

Comment:
We should probably avoid n^2 search for entries to be updated. (As above, for small n, it won't actually be a problem -- so long as `@Transactional` actually means we only issue a couple SQL statements -- but I think the code will also be simpler.)

You could...
* just delete all the old roles and replace them
* structure the API as diffs (additions and removals) instead of overwriting
* create an intermediate `Set` in Java, and test against that instead of doing linear searches

Reply:
I actually talked to @danqrodney and we thought the lists would be small enough that n^2 wasn't too bad. If we are worried about it, we can definitely make a Map ahead of time, and then use that to improve runtime.

Reply:
I'm not really worried about performance, but from a coding standpoint I think using a Map will make things a bit shorter / clearer anyway.